core/lib/zksync_core/src/genesis.rs

```diff
@@ -32,67 +36,144 @@ use zksync_utils::{be_words_to_bytes, bytecode::hash_bytecode, h256_to_u256, u25
+        Self::from_genesis_config(config, base_system_contracts, system_contracts)
+    }
+
     #[cfg(test)]
     pub(crate) fn mock() -> Self {
-        use zksync_types::system_contracts::get_system_smart_contracts;
-
         Self {
-            first_validator: Address::repeat_byte(0x01),
-            protocol_version: ProtocolVersionId::latest(),
             base_system_contracts: BaseSystemContracts::load_from_disk(),
             system_contracts: get_system_smart_contracts(),
-            first_l1_verifier_config: L1VerifierConfig::default(),
+            config: mock_genesis_config(),
         }
     }
-}
 
-pub async fn ensure_genesis_state(
-    storage: &mut StorageProcessor<'_>,
-    zksync_chain_id: L2ChainId,
-    genesis_params: &GenesisParams,
-) -> anyhow::Result<H256> {
-    let mut transaction = storage.start_transaction().await?;
+    pub fn protocol_version(&self) -> ProtocolVersionId {
+        // It's impossible to instantiate Genesis params with wrong protocol version
+        self.config.protocol_version.try_into().expect(
+            "Protocol version must be correctly \
+        initialized for genesis",
+        )
+    }
+}
 
-    // return if genesis block was already processed
-    if !transaction.blocks_dal().is_genesis_needed().await? {
-        tracing::debug!("genesis is not needed!");
-        return transaction
-            .blocks_dal()
-            .get_l1_batch_state_root(L1BatchNumber(0))
-            .await
-            .context("failed fetching state root hash for genesis L1 batch")?
-            .context("genesis L1 batch hash is empty");
+#[cfg(test)]
+pub fn mock_genesis_config() -> GenesisConfig {
+    use zksync_types::L1ChainId;
+
+    let base_system_contracts_hashes = BaseSystemContracts::load_from_disk().hashes();
+    let first_l1_verifier_config = L1VerifierConfig::default();
+
+    GenesisConfig {
+        protocol_version: ProtocolVersionId::latest() as u16,
+        genesis_root_hash: Default::default(),
+        rollup_last_leaf_index: 26,
+        genesis_commitment: Default::default(),
+        bootloader_hash: base_system_contracts_hashes.bootloader,
+        default_aa_hash: base_system_contracts_hashes.default_aa,
+        fee_account: Address::repeat_byte(0x01),
+        l1_chain_id: L1ChainId(9),
+        l2_chain_id: L2ChainId::default(),
+        recursion_node_level_vk_hash: first_l1_verifier_config.params.recursion_node_level_vk_hash,
+        recursion_leaf_level_vk_hash: first_l1_verifier_config.params.recursion_leaf_level_vk_hash,
+        recursion_circuits_set_vks_hash: first_l1_verifier_config
+            .params
+            .recursion_circuits_set_vks_hash,
+        recursion_scheduler_level_vk_hash: first_l1_verifier_config
+            .recursion_scheduler_level_vk_hash,
     }
+}
 
+// This function is dangerous,
+// it doesn't enforce genesis correctness
+// Please, always use `ensure_genesis_state` instead
+pub async fn ensure_genesis_state_unchecked(
```

Comment:
I think the notion of the "dangerous function" comes from a fact that it isn't properly named.
It should be named `insert_genesis_batch` and then there will be no expectations for this function to ensure anything.